File: triage_cli/patch.py
"""
Patch application with safety checks.

Handles:
- Creating git branches before applying
- Validating patches apply cleanly
- Limiting scope of changes
"""

import subprocess
import tempfile
from pathlib import Path
from typing import Optional

from .models.base import Patch


class PatchApplicator:
    """Apply patches with safety checks."""

    def __init__(self, max_files: int = 5, allowed_severities: set = None):
        """
        Initialize patch applicator.

        Args:
            max_files: Maximum files to modify per patch session
            allowed_severities: Only apply patches for these severities (default: S0, S1)
        """
        self.max_files = max_files
        self.allowed_severities = allowed_severities or {'S0', 'S1'}
        self.applied_files = set()

    def apply_patches(
        self,
        patches: list[Patch],
        create_branch: bool = True,
        branch_name: Optional[str] = None
    ) -> int:
        """
        Apply patches to the repository.

        Args:
            patches: List of patches to apply
            create_branch: Whether to create a new branch first
            branch_name: Name for the branch (default: triage/<timestamp>)

        Returns:
            Number of patches successfully applied
        """
        if not patches:
            return 0

        # Create branch if requested
        if create_branch:
            if not self._create_branch(branch_name):
                print("Warning: Failed to create branch, applying to current branch")

        applied = 0
        for patch in patches:
            # Check file limit
            if len(self.applied_files) >= self.max_files:
                print(f"Reached max file limit ({self.max_files}), stopping")
                break

            # Try to apply
            if self._apply_single_patch(patch):
                applied += 1
                self.applied_files.add(patch.path)

        return applied

    def _create_branch(self, branch_name: Optional[str] = None) -> bool:
        """Create a new git branch."""
        if not branch_name:
            from datetime import datetime
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            branch_name = f"triage/{timestamp}"

        try:
            result = subprocess.run(
                ['git', 'checkout', '-b', branch_name],
                capture_output=True,
                text=True,
                timeout=10
            )
            if result.returncode == 0:
                print(f"Created branch: {branch_name}")
                return True
            else:
                print(f"Failed to create branch: {result.stderr}")
                return False
        except Exception as e:
            print(f"Error creating branch: {e}")
            return False

    def _apply_single_patch(self, patch: Patch) -> bool:
        """Apply a single patch."""
        # Validate patch format
        if not self._is_valid_patch(patch):
            print(f"Invalid patch format for {patch.path}")
            return False

        # Check if it applies cleanly (dry run)
        if not self._patch_applies_cleanly(patch):
            print(f"Patch does not apply cleanly to {patch.path}")
            return False

        # Actually apply
        return self._do_apply_patch(patch)

    def _is_valid_patch(self, patch: Patch) -> bool:
        """Validate patch is a valid unified diff."""
        if not patch.diff:
            return False

        # Check for unified diff markers
        lines = patch.diff.split('\n')
        has_minus = any(line.startswith('---') or line.startswith('-') for line in lines)
        has_plus = any(line.startswith('+++') or line.startswith('+') for line in lines)
        has_hunk = any(line.startswith('@@') for line in lines)

        return has_minus and has_plus

    def _patch_applies_cleanly(self, patch: Patch) -> bool:
        """Check if patch applies cleanly (dry run)."""
        try:
            with tempfile.NamedTemporaryFile(mode='w', suffix='.patch', delete=False) as f:
                f.write(patch.diff)
                patch_file = f.name

            result = subprocess.run(
                ['patch', '--dry-run', '-p1', '-i', patch_file],
                capture_output=True,
                text=True,
                timeout=10
            )

            import os
            os.unlink(patch_file)

            return result.returncode == 0

        except Exception as e:
            print(f"Error checking patch: {e}")
            return False

    def _do_apply_patch(self, patch: Patch) -> bool:
        """Actually apply the patch."""
        try:
            with tempfile.NamedTemporaryFile(mode='w', suffix='.patch', delete=False) as f:
                f.write(patch.diff)
                patch_file = f.name

            result = subprocess.run(
                ['patch', '-p1', '-i', patch_file],
                capture_output=True,
                text=True,
                timeout=10
            )

            import os
            os.unlink(patch_file)

            if result.returncode == 0:
                print(f"Applied patch to {patch.path}")
                return True
            else:
                print(f"Failed to apply patch: {result.stderr}")
                return False

        except Exception as e:
            print(f"Error applying patch: {e}")
            return False

    def show_patches(self, patches: list[Patch]) -> str:
        """Show patches without applying (dry run output)."""
        lines = []
        for i, patch in enumerate(patches, 1):
            lines.append(f"=== Patch {i}: {patch.path} ===")
            lines.append(f"Model: {patch.model}")
            lines.append(f"Description: {patch.description}")
            lines.append("")
            lines.append(patch.diff)
            lines.append("")

        return "\n".join(lines)

File: src/patch.ts
/**
 * Patch application with safety checks.
 *
 * Handles:
 * - Creating git branches before applying
 * - Validating patches apply cleanly (dry-run)
 * - Limiting scope of changes
 */

import { execSync } from 'node:child_process';
import { writeFileSync, unlinkSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { randomBytes } from 'node:crypto';
import type { Patch } from './types.js';

// ---------------------------------------------------------------------------
// PatchApplicator
// ---------------------------------------------------------------------------

export class PatchApplicator {
  private maxFiles: number;
  private allowedSeverities: Set<string>;
  private appliedFiles: Set<string>;

  constructor(
    maxFiles = 5,
    allowedSeverities: Set<string> = new Set(['S0', 'S1']),
  ) {
    this.maxFiles = maxFiles;
    this.allowedSeverities = allowedSeverities;
    this.appliedFiles = new Set();
  }

  /**
   * Apply patches to the repository.
   *
   * Returns the number of patches successfully applied.
   */
  applyPatches(
    patches: Patch[],
    createBranch = true,
    branchName?: string,
  ): number {
    if (patches.length === 0) return 0;

    // Create branch if requested — fail-fast to protect current branch
    if (createBranch) {
      if (!this._createBranch(branchName)) {
        console.error('Error: Failed to create branch. Aborting patch application to protect current branch.');
        return 0;
      }
    }

    let applied = 0;

    for (const patch of patches) {
      // Check file limit
      if (this.appliedFiles.size >= this.maxFiles) {
        console.log(`Reached max file limit (${this.maxFiles}), stopping`);
        break;
      }

      if (this._applySinglePatch(patch)) {
        applied++;
        this.appliedFiles.add(patch.path);
      }
    }

    return applied;
  }

  /** Create a new git branch before applying patches. */
  _createBranch(branchName?: string): boolean {
    if (!branchName) {
      const now = new Date();
      const ts =
        now.getFullYear().toString() +
        String(now.getMonth() + 1).padStart(2, '0') +
        String(now.getDate()).padStart(2, '0') +
        '_' +
        String(now.getHours()).padStart(2, '0') +
        String(now.getMinutes()).padStart(2, '0') +
        String(now.getSeconds()).padStart(2, '0');
      branchName = `triage/${ts}`;
    }

    try {
      execSync(`git checkout -b ${branchName}`, {
        stdio: 'pipe',
        timeout: 10_000,
      });
      console.log(`Created branch: ${branchName}`);
      return true;
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      console.log(`Failed to create branch: ${msg}`);
      return false;
    }
  }

  /** Apply a single patch with validation. */
  _applySinglePatch(patch: Patch): boolean {
    if (!this._isValidPatch(patch)) {
      console.log(`Invalid patch format for ${patch.path}`);
      return false;
    }

    if (!this._patchAppliesCleanly(patch)) {
      console.log(`Patch does not apply cleanly to ${patch.path}`);
      return false;
    }

    return this._doApplyPatch(patch);
  }

  /** Validate that the patch is a valid unified diff. */
  _isValidPatch(patch: Patch): boolean {
    if (!patch.diff) return false;

    const lines = patch.diff.split('\n');
    const hasMinus = lines.some(
      (l) => l.startsWith('---') || l.startsWith('-'),
    );
    const hasPlus = lines.some(
      (l) => l.startsWith('+++') || l.startsWith('+'),
    );
    const hasHunk = lines.some((l) => l.startsWith('@@'));

    return hasMinus && hasPlus && hasHunk;
  }

  /** Check if the patch applies cleanly via `patch --dry-run`. */
  _patchAppliesCleanly(patch: Patch): boolean {
    const tmpFile = this._writeTempPatch(patch.diff);
    if (!tmpFile) return false;

    try {
      execSync(`patch --dry-run -p1 -i ${tmpFile}`, {
        stdio: 'pipe',
        timeout: 10_000,
      });
      return true;
    } catch {
      return false;
    } finally {
      this._removeTempFile(tmpFile);
    }
  }

  /** Apply the patch for real via `patch -p1`. */
  _doApplyPatch(patch: Patch): boolean {
    const tmpFile = this._writeTempPatch(patch.diff);
    if (!tmpFile) return false;

    try {
      execSync(`patch -p1 -i ${tmpFile}`, {
        stdio: 'pipe',
        timeout: 10_000,
      });
      console.log(`Applied patch to ${patch.path}`);
      return true;
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      console.log(`Failed to apply patch: ${msg}`);
      return false;
    } finally {
      this._removeTempFile(tmpFile);
    }
  }

  /** Write patch content to a temp file and return its path. */
  private _writeTempPatch(diff: string): string | null {
    const name = join(
      tmpdir(),
      `triage-${randomBytes(6).toString('hex')}.patch`,
    );
    try {
      writeFileSync(name, diff, 'utf-8');
      return name;
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      console.log(`Error writing temp patch: ${msg}`);
      return null;
    }
  }

  /** Remove a temp file, ignoring errors. */
  private _removeTempFile(path: string): void {
    try {
      unlinkSync(path);
    } catch {
      // ignore
    }
  }

  /**
   * Show patches without applying (dry-run display).
   */
  showPatches(patches: Patch[]): string {
    const lines: string[] = [];

    for (let i = 0; i < patches.length; i++) {
      const patch = patches[i];
      lines.push(`=== Patch ${i + 1}: ${patch.path} ===`);
      lines.push(`Model: ${patch.model}`);
      lines.push(`Description: ${patch.description}`);
      lines.push('');
      lines.push(patch.diff);
      lines.push('');
    }

    return lines.join('\n');
  }
}
